File: cpp/facade/facade.cpp
#include <iostream>
#include <memory>

class IChair {
public:
    virtual void sit_on() = 0;
};

class NormalChair : public IChair {
    void sit_on() override {
        std::cout << "Sitting on the chair!" << std::endl;
    }
};

class ITable {
public:
    virtual void eat_from() = 0;
};

class NormalTable : public ITable {
    void eat_from() override {
        std::cout << "Eating from the table!" << std::endl;
    }
};


class KitchenFacade {
public:
    KitchenFacade(std::shared_ptr<ITable> table, std::shared_ptr<IChair> chair) : _table(std::move(table)),
                                                                                  _chair(std::move(chair)) {
    }

    void eat() {
        _chair->sit_on();
        _table->eat_from();
    }

private:
    std::shared_ptr<ITable> _table;
    std::shared_ptr<IChair> _chair;
};


int main() {
    KitchenFacade facade(std::make_shared<NormalTable>(), std::make_shared<NormalChair>());
    facade.eat();
    return 0;
}

File: python/facade/facade.py
from abc import ABC


class IChair(ABC):
    def sit_on(self):
        pass


class ITable(ABC):
    def eat_from(self):
        pass


class NormalChair(IChair):
    def sit_on(self):
        print("Sitting on the chair!")


class NormalTable(ITable):

    def eat_from(self):
        print("Eating from the table!")


class KitchenFacade:
    def __init__(self, table: ITable, chair: IChair):
        self._table = table
        self._chair = chair

    def eat(self):
        self._chair.sit_on()
        self._table.eat_from()


def main():
    facade = KitchenFacade(NormalTable(), NormalChair())
    facade.eat()


if __name__ == '__main__':
    main()
